tmd/areas/create_area_weights.py

```diff
@@ -204,66 +220,79 @@ def create_area_weights_file(area: str, write_file: bool = True):
+    #         article on the augmented Lagrangian method:
+    #         https://en.wikipedia.org/wiki/Augmented_Lagrangian_method
+    #         Instead of applying the multiple to the constraint functions,
+    #         here we apply the inverse of the multiple (which is called
+    #         the REGULARIZATION_LAMBDA) to the sum of the squared weight
+    #         ratio deviations from one, which is logically equivalent.
+    A_dense = (target_matrix * wght_us[:, np.newaxis]).T
+    A = BCOO.from_scipy_sparse(csr_matrix(A_dense))  # A is JAX sparse matrix
+    b = target_array
     print(
-        f"OPTIMIZE_RATIOS= {OPTIMIZE_RATIOS}"
-        f"   target_matrix.shape= {target_matrix.shape}\n"
-        f"OPTIMIZE_LO_BOUND_RATIO= {OPTIMIZE_LO_BOUND_RATIO:.1f}\n"
-        f"OPTIMIZE_HI_BOUND_RATIO= {OPTIMIZE_HI_BOUND_RATIO:.1f}"
+        f"OPTIMIZE_RATIOS: target_matrix.shape= {target_matrix.shape}\n"
+        f"REGULARIZATION_LAMBDA= {REGULARIZATION_LAMBDA:e}"
     )
     time0 = time.time()
-    res = lsq_linear(
-        tar_matrix,
-        target_array,
-        bounds=(lob, hib),
-        method="bvls",
-        tol=OPTIMIZE_FTOL,
-        lsq_solver="exact",
-        lsmr_tol=None,
-        max_iter=OPTIMIZE_MAXITER,
-        verbose=OPTIMIZE_VERBOSE,
+    res = minimize(
+        fun=JIT_FVAL_AND_GRAD,  # objective function and its gradient
+        x0=np.ones(num_weights),  # initial wght_ratio guess
+        jac=True,  # use gradient from JIT_FVAL_AND_GRAD function
+        args=(A, b, REGULARIZATION_LAMBDA),  # fixed arguments
+        method="L-BFGS-B",  # use L-BFGS-B algorithm
+        bounds=Bounds(0.0, np.inf),  # consider only non-negative weight ratios
+        options={
+            "maxiter": OPTIMIZE_MAXITER,
+            "ftol": OPTIMIZE_FTOL,
+            "gtol": OPTIMIZE_GTOL,
+            "iprint": OPTIMIZE_IPRINT,
+        },
     )
     time1 = time.time()
     res_summary = (
-        f">>> scipy.lsq_linear execution: {(time1-time0):.1f} secs"
+        f">>> optimization execution time: {(time1-time0):.1f} secs"
         f"  iterations={res.nit}  success={res.success}\n"
         f">>> message: {res.message}\n"
-        f">>> lsq_linear optimized loss value: {res.cost:.9e}"
+        f">>> L-BFGS-B optimized loss value: {res.fun:.9e}"
     )
     print(res_summary)
     if OPTIMIZE_RESULTS:
-        print(">>> scipy.lsq_linear full results:\n", res)
-    if OPTIMIZE_RATIOS:
-        wght_ratio = res.x
-        wght_area = res.x * wght_us
-    else:
-        wght_ratio = res.x / wght_us
-        wght_area = res.x
-    loss = loss_function_value(wght_area, target_matrix, target_array)
-    print(f"AREA-OPTIMIZED_LOSS_FUNCTION_VALUE= {loss:.9e}")
+        print(">>> full optimization results:\n", res)
+    wght_ratio = res.x
+    wght_area = res.x * wght_us
+    loss = target_loss_function_value(wght_area, target_matrix, target_array)
+    print(f"AREA-OPTIMIZED_TARGET_LOSS_FUNCTION_VALUE= {loss:.9e}")
```

Comment:
I do think having quantiles (or bins) of proportionate differences from targets is important. In some real-world problems we won't be able to hit all targets, or won't like the results when we do so (e.g., can only hit targets if we make lambda near-zero and accept huge x ratios. We'll also want diagnostic capability to see which differences are far from zero. The pre-optimization target printout is useful for that purpose but again, if we have 500 targets that will be a lot to wade through. It might be worth having an if EXAMINE_TARGETS sort of block.

We also may want some pre-optimization examination of the problem, but it could actually occur outside of the python code; I'll give examples of what I mean if we get into this more deeply under contract.

Reply:
@donboyd5 said:
```
I do think having quantiles [...] of proportionate differences from targets is important.
```

Working on this now.

Reply:
@donboyd5,  I added a bin distribution of target actual-to-expect ratios.

Reply:
@martinholmer That's great - thank you.